<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Normalize raw count data — normalize • DIANE</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Normalize raw count data — normalize"><meta property="og:description" content="The function corrects for different sequencing depths bewteen samples.
It relies on the TCC package, to build a TCC-class object containing the raw counts 
and the conditions for each sample. The function calcNormFactors is then applied, 
and uses the method chosen by the user. It can, weather or not, proceed to a first step or 
removing potentially differentially expressed genes to have less biased normalisation
factors in the second normalization step. It returns a TCC object,
with an element norm_factors containing the computed normalization factors."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">DIANE</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/DIANE.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/DIANE_Input_data.html">Data upload</a>
    </li>
    <li>
      <a href="../articles/DIANE_Programming_Interface.html">DIANE Programming Interface</a>
    </li>
    <li>
      <a href="../articles/DIANE_normalize.html">Normalization</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/OceaneCsn/DIANE/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Normalize raw count data</h1>
    <small class="dont-index">Source: <a href="https://github.com/OceaneCsn/DIANE/blob/HEAD/R/fct_normalisation.R" class="external-link"><code>R/fct_normalisation.R</code></a></small>
    <div class="hidden name"><code>normalize.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The function corrects for different sequencing depths bewteen samples.
It relies on the TCC package, to build a TCC-class object containing the raw counts 
and the conditions for each sample. The function calcNormFactors is then applied, 
and uses the method chosen by the user. It can, weather or not, proceed to a first step or 
removing potentially differentially expressed genes to have less biased normalisation
factors in the second normalization step. It returns a TCC object,
with an element norm_factors containing the computed normalization factors.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">normalize</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  conditions <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split_fixed</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, <span class="st">"_"</span>, <span class="fl">2</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>,</span>
<span>  norm_method <span class="op">=</span> <span class="st">"tmm"</span>,</span>
<span>  deg_method <span class="op">=</span> <span class="st">"deseq2"</span>,</span>
<span>  fdr <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  iteration <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data</dt>
<dd><p>raw counts to be normalized (data frame or matrix), with genes as rownames and
conditions as columns.</p></dd>


<dt>conditions</dt>
<dd><p>condition of each column of the data argument. Default is
all the conditions in the experiment. (as defined by the underscore prefixes).</p></dd>


<dt>norm_method</dt>
<dd><p>method used for normalization, between tmm or deseq2</p></dd>


<dt>deg_method</dt>
<dd><p>method used for DEGs detection if chosen, between edgeR ou deseq2</p></dd>


<dt>fdr</dt>
<dd><p>pvalue threshold for adjusted pvalues for DEGs detection if chosen</p></dd>


<dt>iteration</dt>
<dd><p>weather or not to perform a prior removal of DEGs (TRUE or FALSE)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a TCC-Class object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Filtering low counts is highly recommended after normalization, 
consider using the <code><a href="filter_low_counts.html">DIANE::filter_low_counts</a></code>
function just after this function.</p>
<p>You can get the normalized expression matrix with <code>TCC::getNormalizedData(tcc)</code>,
tcc being the result of <code>DIANE::normalize()</code> or <code><a href="filter_low_counts.html">DIANE::filter_low_counts()</a></code></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"abiotic_stresses"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tcc_object</span> <span class="op">&lt;-</span> <span class="fu">DIANE</span><span class="fu">::</span><span class="fu">normalize</span><span class="op">(</span><span class="va">abiotic_stresses</span><span class="op">$</span><span class="va">raw_counts</span>, </span></span>
<span class="r-in"><span><span class="va">abiotic_stresses</span><span class="op">$</span><span class="va">conditions</span>, iteration <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://oceanecsn.github.io/" class="external-link">Océane Cassan</a>, Alexandre Soriano.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

